Using relay.node() in tests causes DeprecationWarning

## Describe the Bug
When relay.node is included in a query schema the test fails with an obsure DeprecationWarning bug. The following is a contrived example to demonstrate the problem.

```python
@pytest.mark.asyncio
async def test_node_query():
    @strawberry.type
    class Query:
        @strawberry.field
        def hello(self) -> str:
            return "Hello World"

        node: relay.Node = relay.node()

    schema = strawberry.Schema(
        query=Query,
    )
    query = """
        query Query {
            hello
        }
    """
    result = await schema.execute(query)
```
Generates an error:
```
=================================== FAILURES ===================================
_______________________________ test_node_query ________________________________

self = <GraphQLObjectType 'Query'>

    @cached_property
    def fields(self) -> GraphQLFieldMap:
        """Get provided fields, wrapping them as GraphQLFields if needed."""
        try:
>           fields = resolve_thunk(self._fields)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

.venv/lib/python3.12/site-packages/graphql/type/definition.py:816: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
backend/.venv/lib/python3.12/site-packages/graphql/type/definition.py:300: in resolve_thunk
    return thunk() if callable(thunk) else thunk
           ^^^^^^^
.venv/lib/python3.12/site-packages/strawberry/schema/schema_converter.py:673: in <lambda>
    fields=lambda: self.get_graphql_fields(object_type),
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/strawberry/schema/schema_converter.py:477: in get_graphql_fields
    return _get_thunk_mapping(
.venv/lib/python3.12/site-packages/strawberry/schema/schema_converter.py:145: in _get_thunk_mapping
    thunk_mapping[name_converter(field)] = field_converter(
backend/.venv/lib/python3.12/site-packages/strawberry/schema/schema_converter.py:414: in from_field
    resolver = self.from_resolver(field)
               ^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/strawberry/schema/schema_converter.py:772: in from_resolver
    _get_result_with_extensions = wrap_field_extensions()
                                  ^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/strawberry/schema/schema_converter.py:723: in wrap_field_extensions
    extension.apply(field)
.venv/lib/python3.12/site-packages/strawberry/relay/fields.py:65: in apply
    field.base_resolver = StrawberryResolver(resolver, type_override=field.type)
    ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/strawberry/types/field.py:293: in base_resolver
    _ = resolver.arguments
        ^^^^^^^^^^^^^^^^^^
../../../.local/share/uv/python/cpython-3.12.9-macos-aarch64-none/lib/python3.12/functools.py:998: in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/strawberry/types/fields/resolver.py:244: in arguments
    root_parameter = self.reserved_parameters.get(ROOT_PARAMSPEC)
                     ^^^^^^^^^^^^^^^^^^^^^^^^
../../../.local/share/uv/python/cpython-3.12.9-macos-aarch64-none/lib/python3.12/functools.py:998: in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.12/site-packages/strawberry/types/fields/resolver.py:239: in reserved_parameters
    return {spec: spec.find(parameters, self) for spec in self.RESERVED_PARAMSPEC}
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ReservedType(name='info', type=<class 'strawberry.types.info.Info'>)
parameters = (<Parameter "info: 'Info'">, <Parameter "id: "Annotated[GlobalID, argument(description='The ID of the object.')]"">)
resolver = <strawberry.types.fields.resolver.StrawberryResolver object at 0x353fe1a00>

    def find(
        self,
        parameters: tuple[inspect.Parameter, ...],
        resolver: StrawberryResolver[Any],
    ) -> Optional[inspect.Parameter]:
        # Go through all the types even after we've found one so we can
        # give a helpful error message if someone uses the type more than once.
        type_parameters = []
        for parameter in parameters:
            annotation = resolver.strawberry_annotations[parameter]
            if isinstance(annotation, StrawberryAnnotation):
                try:
                    evaled_annotation = annotation.evaluate()
                except NameError:
                    # If this is a strawberry.Parent using ForwardRef, we will fail to
                    # evaluate at this moment, but at least knowing that it is a reserved
                    # type is enough for now
                    # We might want to revisit this in the future, maybe by postponing
                    # this check to when the schema is actually being created
                    evaled_annotation = resolve_parent_forward_arg(
                        annotation.annotation
                    )
    
                if self.is_reserved_type(evaled_annotation):
                    type_parameters.append(parameter)
    
        if len(type_parameters) > 1:
            raise ConflictingArgumentsError(
                resolver, [parameter.name for parameter in type_parameters]
            )
    
        if type_parameters:
            return type_parameters[0]
    
        # Fallback to matching by name
        if not self.name:
            return None
        reserved_name = ReservedName(name=self.name).find(parameters, resolver)
        if reserved_name:
            warning = DeprecationWarning(
                f"Argument name-based matching of '{self.name}' is deprecated and will "
                "be removed in v1.0. Ensure that reserved arguments are annotated "
                "their respective types (i.e. use value: 'DirectiveValue[str]' instead "
                "of 'value: str' and 'info: Info' instead of a plain 'info')."
            )
>           warnings.warn(warning, stacklevel=3)
E           DeprecationWarning: Argument name-based matching of 'info' is deprecated and will be removed in v1.0. Ensure that reserved arguments are annotated their respective types (i.e. use value: 'DirectiveValue[str]' instead of 'value: str' and 'info: Info' instead of a plain 'info').

.venv/lib/python3.12/site-packages/strawberry/types/fields/resolver.py:152: DeprecationWarning

The above exception was the direct cause of the following exception:

    @pytest.mark.asyncio
    async def test_node_query():
        @strawberry.type
        class Query:
            @strawberry.field
            def hello(self) -> str:
                return "Hello World"
    
            node: relay.Node = relay.node()
    
>       schema = strawberry.Schema(
            query=Query,
        )
```
## System Information

 - Operating system:
 - Python version: 3.12
 - Strawberry version (if applicable): "strawberry-graphql>=0.282.0"

## Additional Context

Removing the relay.node from the above test allows it to pass

--- a/RELEASE.md
+++ b/RELEASE.md
@@ -1,0 +1,12 @@
+Release type: patch
+
+Fixed a bug where using `relay.node()` or `relay.connection()` would raise a
+`DeprecationWarning` about "Argument name-based matching of 'info'", even though
+the `info` parameter was correctly annotated.
+
+This also fixes the same warning when using Python 3.12+ type alias syntax for
+`Info`, such as `type MyInfo = strawberry.Info[Context, None]`.
+
+The fix recognizes `strawberry.Info` and `strawberry.Parent` annotations (with
+or without generic parameters) when they appear as string forward references.
+Only fully qualified names are matched to avoid confusion with user-defined types.

--- a/strawberry/field_extensions/input_mutation.py
+++ b/strawberry/field_extensions/input_mutation.py
@@ -1,9 +1,6 @@
 from __future__ import annotations
 
-from typing import (
-    TYPE_CHECKING,
-    Any,
-)
+from typing import Any
 
 import strawberry
 from strawberry.annotation import StrawberryAnnotation
@@ -15,9 +12,6 @@ from strawberry.extensions.field_extension import (
 from strawberry.types.arguments import StrawberryArgument
 from strawberry.types.field import StrawberryField
 from strawberry.utils.str_converters import capitalize_first, to_camel_case
-
-if TYPE_CHECKING:
-    from strawberry.types.info import Info
 
 
 class InputMutationExtension(FieldExtension):
@@ -64,7 +58,7 @@ class InputMutationExtension(FieldExtension):
         self,
         next_: SyncExtensionResolver,
         source: Any,
-        info: Info,
+        info: strawberry.Info,
         **kwargs: Any,
     ) -> Any:
         input_args = kwargs.pop("input")
@@ -79,7 +73,7 @@ class InputMutationExtension(FieldExtension):
         self,
         next_: AsyncExtensionResolver,
         source: Any,
-        info: Info,
+        info: strawberry.Info,
         **kwargs: Any,
     ) -> Any:
         input_args = kwargs.pop("input")

--- a/strawberry/relay/fields.py
+++ b/strawberry/relay/fields.py
@@ -49,8 +49,8 @@ from .types import Connection, GlobalID, Node
 if TYPE_CHECKING:
     from typing import Literal
 
+    import strawberry
     from strawberry.permission import BasePermission
-    from strawberry.types.info import Info
 
 
 class NodeExtension(FieldExtension):
@@ -65,12 +65,20 @@ class NodeExtension(FieldExtension):
         field.base_resolver = StrawberryResolver(resolver, type_override=field.type)
 
     def resolve(
-        self, next_: SyncExtensionResolver, source: Any, info: Info, **kwargs: Any
+        self,
+        next_: SyncExtensionResolver,
+        source: Any,
+        info: strawberry.Info,
+        **kwargs: Any,
     ) -> Any:
         return next_(source, info, **kwargs)
 
     async def resolve_async(
-        self, next_: SyncExtensionResolver, source: Any, info: Info, **kwargs: Any
+        self,
+        next_: SyncExtensionResolver,
+        source: Any,
+        info: strawberry.Info,
+        **kwargs: Any,
     ) -> Any:
         retval = next_(source, info, **kwargs)
         # If the resolve_nodes method is not async, retval will not actually
@@ -81,12 +89,12 @@ class NodeExtension(FieldExtension):
 
     def get_node_resolver(
         self, field: StrawberryField
-    ) -> Callable[[Info, GlobalID], Node | None | Awaitable[Node | None]]:
+    ) -> Callable[[strawberry.Info, GlobalID], Node | None | Awaitable[Node | None]]:
         type_ = field.type
         is_optional = isinstance(type_, StrawberryOptional)
 
         def resolver(
-            info: Info,
+            info: strawberry.Info,
             id: Annotated[GlobalID, argument(description="The ID of the object.")],
         ) -> Node | None | Awaitable[Node | None]:
             node_type = id.resolve_type(info)
@@ -114,13 +122,15 @@ class NodeExtension(FieldExtension):
 
     def get_node_list_resolver(
         self, field: StrawberryField
-    ) -> Callable[[Info, list[GlobalID]], list[Node] | Awaitable[list[Node]]]:
+    ) -> Callable[
+        [strawberry.Info, list[GlobalID]], list[Node] | Awaitable[list[Node]]
+    ]:
         type_ = field.type
         assert isinstance(type_, StrawberryList)
         is_optional = isinstance(type_.of_type, StrawberryOptional)
 
         def resolver(
-            info: Info,
+            info: strawberry.Info,
             ids: Annotated[
                 list[GlobalID], argument(description="The IDs of the objects.")
             ],
@@ -306,7 +316,7 @@ class ConnectionExtension(FieldExtension):
         self,
         next_: SyncExtensionResolver,
         source: Any,
-        info: Info,
+        info: strawberry.Info,
         *,
         before: str | None = None,
         after: str | None = None,
@@ -329,7 +339,7 @@ class ConnectionExtension(FieldExtension):
         self,
         next_: AsyncExtensionResolver,
         source: Any,
-        info: Info,
+        info: strawberry.Info,
         *,
         before: str | None = None,
         after: str | None = None,

--- a/strawberry/relay/utils.py
+++ b/strawberry/relay/utils.py
@@ -10,7 +10,7 @@ from strawberry.types.base import StrawberryObjectDefinition
 from strawberry.types.nodes import InlineFragment
 
 if TYPE_CHECKING:
-    from strawberry.types.info import Info
+    import strawberry
 
 
 def from_base64(value: str) -> tuple[str, str]:
@@ -71,7 +71,7 @@ def to_base64(type_: str | type | StrawberryObjectDefinition, node_id: Any) -> s
     return base64.b64encode(f"{type_name}:{node_id}".encode()).decode()
 
 
-def should_resolve_list_connection_edges(info: Info) -> bool:
+def should_resolve_list_connection_edges(info: strawberry.Info) -> bool:
     """Check if the user requested to resolve the `edges` field of a connection.
 
     Args:
@@ -113,7 +113,7 @@ class SliceMetadata:
     @classmethod
     def from_arguments(
         cls,
-        info: Info,
+        info: strawberry.Info,
         *,
         before: str | None = None,
         after: str | None = None,

--- a/strawberry/types/fields/resolver.py
+++ b/strawberry/types/fields/resolver.py
@@ -10,6 +10,7 @@ from typing import (
     TYPE_CHECKING,
     Annotated,
     Any,
+    ForwardRef,
     Generic,
     NamedTuple,
     TypeVar,
@@ -97,10 +98,14 @@ class ReservedType(NamedTuple):
 
     To preserve backwards-comaptibility, if an annotation was defined but does not match
     :attr:`type`, then the name is used as a fallback if available.
+
+    :attr:`alias` is the public name exposed in the `strawberry` module, if it differs
+    from the internal class name (e.g. "Parent" for StrawberryParent).
     """
 
     name: str | None
     type: type
+    alias: str | None = None
 
     def find(
         self,
@@ -114,16 +119,18 @@ class ReservedType(NamedTuple):
             annotation = resolver.strawberry_annotations[parameter]
             if isinstance(annotation, StrawberryAnnotation):
                 try:
-                    evaled_annotation = annotation.evaluate()
+                    evaled_annotation: Any = annotation.evaluate()
                 except NameError:
-                    # If this is a strawberry.Parent using ForwardRef, we will fail to
-                    # evaluate at this moment, but at least knowing that it is a reserved
-                    # type is enough for now
-                    # We might want to revisit this in the future, maybe by postponing
-                    # this check to when the schema is actually being created
-                    evaled_annotation = resolve_parent_forward_arg(
-                        annotation.annotation
-                    )
+                    # If we fail to evaluate, check if the raw annotation string
+                    # matches this reserved type. This handles cases where types
+                    # are imported under TYPE_CHECKING or use forward references
+                    # like "strawberry.Info".
+                    raw = annotation.annotation
+                    if isinstance(raw, str):
+                        evaled_annotation = raw
+                    else:
+                        # Try resolve_parent_forward_arg for Parent types
+                        evaled_annotation = resolve_parent_forward_arg(raw)
 
                 if self.is_reserved_type(evaled_annotation):
                     type_parameters.append(parameter)
@@ -151,24 +158,56 @@ class ReservedType(NamedTuple):
             return reserved_name
         return None
 
-    def is_reserved_type(self, other: builtins.type) -> bool:
+    def is_reserved_type(self, other: builtins.type | str | ForwardRef) -> bool:
+        # Handle string forward references (e.g., "strawberry.Info", "Info[Context, None]")
+        # This occurs when annotations can't be resolved at runtime, such as when
+        # the type is imported under TYPE_CHECKING or uses `from __future__ import annotations`
+        if isinstance(other, str):
+            return self._is_reserved_type_str(other)
+
+        if isinstance(other, ForwardRef):
+            return self._is_reserved_type_str(other.__forward_arg__)
+
+        # Handle TypeAliasType (Python 3.12+ `type X = ...` syntax)
+        # We need to unwrap the alias to get the actual type
+        if hasattr(other, "__value__"):
+            other = other.__value__
+
         origin = cast("type", get_origin(other)) or other
         if origin is Annotated:
             # Handle annotated arguments such as Private[str] and DirectiveValue[str]
             return type_has_annotation(other, self.type)
-        # Handle both concrete and generic types (i.e Info, and Info)
+
+        # Handle both concrete and generic types (i.e. Info, and Info[X, Y])
         return (
             issubclass(origin, self.type)
             if isinstance(origin, type)
             else origin is self.type
         )
 
+    def _is_reserved_type_str(self, annotation: str) -> bool:
+        type_name = self.type.__name__
+        module = self.type.__module__
+
+        base_annotation = annotation.split("[")[0].strip()
+        # Only match qualified names (strawberry.X or full module path)
+        valid_names = {
+            f"strawberry.{type_name}",  # e.g. "strawberry.Info"
+            f"{module}.{type_name}",  # e.g. "strawberry.types.info.Info"
+        }
+
+        # Add alias if provided (e.g. "Parent" for StrawberryParent)
+        if self.alias:
+            valid_names.add(f"strawberry.{self.alias}")
+
+        return base_annotation in valid_names
+
 
 SELF_PARAMSPEC = ReservedNameBoundParameter("self")
 CLS_PARAMSPEC = ReservedNameBoundParameter("cls")
 ROOT_PARAMSPEC = ReservedName("root")
 INFO_PARAMSPEC = ReservedType("info", Info)
-PARENT_PARAMSPEC = ReservedType(name=None, type=StrawberryParent)
+PARENT_PARAMSPEC = ReservedType(name=None, type=StrawberryParent, alias="Parent")
 
 T = TypeVar("T")
 

--- a/tests/fields/test_resolvers.py
+++ b/tests/fields/test_resolvers.py
@@ -1,7 +1,7 @@
 import dataclasses
 import textwrap
 import types
-from typing import Any, ClassVar, no_type_check
+from typing import Any, ClassVar, ForwardRef, no_type_check
 
 import pytest
 
@@ -14,6 +14,7 @@ from strawberry.exceptions import (
 from strawberry.parent import Parent
 from strawberry.scalars import JSON
 from strawberry.types.fields.resolver import (
+    INFO_PARAMSPEC,
     Signature,
     StrawberryResolver,
     UncallableResolverError,
@@ -535,3 +536,75 @@ def test_annotation_using_parent_annotation_but_named_root():
             "nameFromParent": "Using 'root': Strawberry",
         }
     }
+
+
+def test_info_annotation_with_type_alias():
+    """Test that TypeAliasType for Info is properly recognized (issue #4057)."""
+    from typing_extensions import TypeAliasType
+
+    from strawberry.types.info import Info
+
+    # Simulate Python 3.12's `type MyInfo = Info[None, None]`
+    MyInfo = TypeAliasType("MyInfo", Info[None, None])
+
+    @strawberry.type
+    class Query:
+        @strawberry.field
+        def hello(self, info: MyInfo) -> str:
+            return "Hello World"
+
+    # Should not raise DeprecationWarning
+    schema = strawberry.Schema(query=Query)
+    result = schema.execute_sync("query { hello }")
+    assert result.errors is None
+    assert result.data == {"hello": "Hello World"}
+
+
+def test_info_forward_reference_string():
+    """Test that string forward references for Info are properly recognized.
+
+    Only qualified names (strawberry.Info or full module path) are matched
+    to avoid confusion with user-defined types.
+    """
+
+    # Test qualified string patterns that should match Info
+    assert INFO_PARAMSPEC.is_reserved_type("strawberry.Info") is True
+    assert INFO_PARAMSPEC.is_reserved_type("strawberry.Info[Context, None]") is True
+    assert INFO_PARAMSPEC.is_reserved_type("strawberry.types.info.Info") is True
+    assert (
+        INFO_PARAMSPEC.is_reserved_type("strawberry.types.info.Info[Context, None]")
+        is True
+    )
+
+    # Test patterns that should NOT match (unqualified or wrong names)
+    assert INFO_PARAMSPEC.is_reserved_type("Info") is False  # unqualified
+    assert INFO_PARAMSPEC.is_reserved_type("Info[Context, None]") is False
+    assert INFO_PARAMSPEC.is_reserved_type("str") is False
+    assert INFO_PARAMSPEC.is_reserved_type("InfoClass") is False
+    assert INFO_PARAMSPEC.is_reserved_type("MyInfo") is False
+
+
+def test_info_forward_ref_object():
+    """Test that ForwardRef objects for Info are properly recognized.
+
+    Only qualified names (strawberry.Info or full module path) are matched
+    to avoid confusion with user-defined types.
+    """
+
+    from strawberry.types.fields.resolver import INFO_PARAMSPEC
+
+    # Test ForwardRef patterns that should match Info
+    assert INFO_PARAMSPEC.is_reserved_type(ForwardRef("strawberry.Info")) is True
+    assert (
+        INFO_PARAMSPEC.is_reserved_type(ForwardRef("strawberry.Info[Context, None]"))
+        is True
+    )
+    assert (
+        INFO_PARAMSPEC.is_reserved_type(ForwardRef("strawberry.types.info.Info"))
+        is True
+    )
+
+    # Test ForwardRef patterns that should NOT match
+    assert INFO_PARAMSPEC.is_reserved_type(ForwardRef("Info")) is False  # unqualified
+    assert INFO_PARAMSPEC.is_reserved_type(ForwardRef("str")) is False
+    assert INFO_PARAMSPEC.is_reserved_type(ForwardRef("MyInfo")) is False

--- a/tests/python_312/test_generics_schema.py
+++ b/tests/python_312/test_generics_schema.py
@@ -1,6 +1,7 @@
 # ruff: noqa: F821
 
 import sys
+import warnings
 import textwrap
 from enum import Enum
 from typing import Any, Optional, Union
@@ -1220,3 +1221,22 @@ def test_generics_via_anonymous_union():
     ).strip()
 
     assert str(schema) == expected_schema
+
+
+def test_info_type_alias_syntax():
+    type MyInfo = strawberry.Info[None, None]  # type: ignore[valid-type]
+
+    @strawberry.type
+    class Query:
+        @strawberry.field
+        def hello(self, info: MyInfo) -> str:
+            return "Hello World"
+
+    # Should not raise DeprecationWarning about name-based matching
+    with warnings.catch_warnings():
+        warnings.filterwarnings("error", category=DeprecationWarning)
+        schema = strawberry.Schema(query=Query)
+
+    result = schema.execute_sync("query { hello }")
+    assert result.errors is None
+    assert result.data == {"hello": "Hello World"}

--- a/tests/relay/test_fields.py
+++ b/tests/relay/test_fields.py
@@ -1,5 +1,6 @@
 import dataclasses
 import textwrap
+import warnings
 from collections.abc import Iterable
 from typing_extensions import Self
 
@@ -1687,3 +1688,26 @@ def test_correct_model_returned(type_name: str, should_have_name: bool):
         assert result.data["node"]["name"] == "Strawberry"
     else:
         assert "name" not in result.data["node"]
+
+
+def test_relay_node_no_deprecation_warning():
+    @strawberry.type
+    class Query:
+        @strawberry.field
+        def hello(self) -> str:
+            return "Hello World"
+
+        node: relay.Node = relay.node()
+
+    # Should not raise DeprecationWarning about name-based matching
+    with warnings.catch_warnings():
+        warnings.filterwarnings("error", category=DeprecationWarning)
+        schema = strawberry.Schema(query=Query)
+
+    result = schema.execute_sync(
+        """
+        query { hello }
+        """
+    )
+    assert result.errors is None
+    assert result.data == {"hello": "Hello World"}

